### PR TITLE
Document models

### PR DESCRIPTION
**Pull request by SKairinos**
This change is [Reviewable](https://reviewable.io/reviews/ocadotechnology/codeforlife-workspace/97)

### CODE REVIEW

**Review by faucomte97 — COMMENTED** [review]
> Reviewed 6 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/codeforlife-workspace/97)* status: all files reviewed, 8 unresolved discussions (waiting on @SKairinos)

**Review by SKairinos — COMMENTED** [review]
feedback

File: docs/service/backend/MODELS.md [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+TL;DR
+
+- [x] One file per model in the `models` directory.
+- [x] Each model-file follows the naming convention `{model}.py`.
+- [x] Each model is imported from its file in `models/__init__.py`.
+- [x] One test-file per model in the `models` directory.
+- [x] Each model-test-file follows the naming convention `{model}_test.py`.
+- [x] Each model-test-case follows the naming convention `Test{model}`.
+- [x] Each model-test-case inherits `ModelTestCase`.
+- [x] Each model-test-case set their type parameter to the model being tested.
```
[the remark this answers is not in the document]
Comment (SKairinos) [reply docs/service/backend/MODELS.md]: Done.

File: docs/service/backend/MODELS.md [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+TL;DR
+
+- [x] Inherit `WarehouseModel` if storing the model's historical data will provide meaningful insights.
+- [x] Inherit `models.Model` if storing the model's historical data will NOT provide meaningful insights.
+
+---
+
+Determine whether the data stored in a model needs to be warehoused. Data should be warehoused if it can contribute to analyzing users' behavior. By default, new models should be warehoused as it's a rare occurrence that meaningful insights cannot be gained from analyzing its data.
+
+Defining a model as a warehouse-model allows us to sync data from our database to our data warehouse before its deleted from our database. See our [data deletion strategy](https://code-for-life.gitbook.io/code-for-life-dev/data/storage#data-deletion-strategy).
```
[the remark this answers is not in the document]
Comment (SKairinos) [reply docs/service/backend/MODELS.md]: Done.

File: docs/service/backend/MODELS.md [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+```
+
+## Defining Managers
+
+TL;DR
+
+If you're defining a model with a custom manager:
+
+- [x] Define object manager as `Manager` within the model.
+- [x] If the model inherits `WarehouseModel`, its manager should inherit `WarehouseModel.Manger` and set the type variable to a string with the model's name.
```
[the remark this answers is not in the document]
Comment (SKairinos) [reply docs/service/backend/MODELS.md]: Done.

File: docs/service/backend/MODELS.md [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+- [x] Write `objects: Manager = Manger()` below a custom manager.
```
[the remark this answers is not in the document]
Comment (SKairinos) [reply docs/service/backend/MODELS.md]: Done.

File: docs/service/backend/MODELS.md [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+
+If you're defining an abstract model with a custom manager:
+
+- [x] Create a type variable lazily bound to the abstract model with naming convention `Any{model}`.
+- [x] Add generic type parameter of type any model that inherits the abstract model on the manager.
+- [x] Write `objects: Manager[t.Self] = Manger()` below a custom manager.
+
+---
+
+A model's manager should be defined within the model's class as `Manager`. If the model inherits `WarehouseModel`, the manager should inherit `WarehouseModel.Manager`, providing a string of the model's name as the type parameter to inform the manager of the type of model it will be managing. Below the manager's definition, create a class-level attribute on the model called `objects` which has its type set to `Manager` and it's value set to an instance of the `Manager` class.
```
[the remark this answers is not in the document]
Comment (SKairinos) [reply docs/service/backend/MODELS.md]: Done.

**Review by SKairinos — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/codeforlife-workspace/97)* status: 5 of 6 files reviewed, 8 unresolved discussions (waiting on @faucomte97)

___
*[`docs/service/backend/MODELS.md` line 205 at r1](https://reviewable.io/reviews/ocadotechnology/codeforlife-workspace/97#gh-1549970756:-NuyCr-n-Puo7cjmUWkx:b-896fix) ([raw file](https://github.com/ocadotechnology/codeforlife-workspace/blob/4ca856d5f0de8413e5ed1ffa28cfb312cbe00f7d/docs/service/backend/MODELS.md#L205)):*
<details><summary><i>Previously, faucomte97 (Florian Aucomte) wrote…</i></summary><blockquote>

The little lord Jesus 
Lay down his sweet head
</blockquote></details>

Done.

File: docs/service/backend/SERIALIZERS.md [hunk 6]
```diff
@@ -0,0 +1,235 @@
+# Serializers
+
+Below are the conventions we follow when serializing our data.
+
+First, understand how to [serialize a Django model](https://www.django-rest-framework.org/api-guide/serializers/#modelserializer).
+
+## Model Serializers
+
+For each model, we create a file with the naming convention `{model}.py` in the directory `serializers`. Within a model's serializer-file, we create one or more serializers. It's advised to create one serializer per [model-view-set](./VIEWS.md#model-view-sets) action. Each model-serializer should inherit CFL's `ModelSerializer` by default and set the type parameter to the model being serialized.
+
```
[the remark this answers is not in the document]
Comment (SKairinos) [reply docs/service/backend/SERIALIZERS.md:10]: Done.

File: docs/service/backend/VIEWS.md [hunk 7]
```diff
@@ -0,0 +1,298 @@
+# Views
+
+Below are the conventions we follow when viewing our data.
+
+First, understand how to [view a Django model](https://www.django-rest-framework.org/api-guide/viewsets/#modelviewset).
+
+## Model View Sets
+
```
[the remark this answers is not in the document]
Comment (SKairinos) [reply docs/service/backend/VIEWS.md:8]: Done

**Review by faucomte97 — COMMENTED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/codeforlife-workspace/97)* status: all files reviewed, 2 unresolved discussions (waiting on @SKairinos)

**Review by faucomte97 — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/codeforlife-workspace/97#-:-NuyJq1c74PXf58JJedh:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @SKairinos)

**Pull request merged**